lib/auth/session_access.go

```diff
@@ -0,0 +1,265 @@
+	}
+
+	return fn(), nil
+}
+
+func (e *SessionAccessEvaluator) matchesJoin(allow *types.SessionJoinPolicy) bool {
+	if !e.matchesKind(allow.Kinds) {
+		return false
+	}
+
+	for _, requireRole := range e.roles {
+		for _, allowRole := range allow.Roles {
+			matched := glob.Glob(requireRole.GetName(), allowRole)
+
+			if matched {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (e *SessionAccessEvaluator) matchesKind(allow []string) bool {
+	if contains(allow, e.kind) || contains(allow, "*") {
+		return true
+	}
+
+	return false
+}
+
+// CanJoin checks if a given user is eligible to join a session.
+func (e *SessionAccessEvaluator) CanJoin(user SessionAccessContext) []types.SessionParticipantMode {
+	if !e.isPoisoned() {
+		return []types.SessionParticipantMode{
+			types.SessionObserverMode,
+			types.SessionModeratorMode,
+			types.SessionPeerMode,
+		}
+	}
+
+	var modes []types.SessionParticipantMode
+
+	for _, allowPolicy := range getAllowPolicies(user) {
+		if e.matchesJoin(allowPolicy) {
+			for _, modeString := range allowPolicy.Modes {
+				mode := types.SessionParticipantMode(modeString)
+				if !SliceContainsMode(modes, mode) {
+					modes = append(modes, mode)
+				}
+			}
+		}
+	}
+
+	return modes
+}
+
+func SliceContainsMode(s []types.SessionParticipantMode, e types.SessionParticipantMode) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+// PolicyOptions is a set of settings for the session determined by the matched require policy.
+type PolicyOptions struct {
+	TerminateOnLeave bool
+}
+
+// FulfilledFor checks if a given session may run with a list of participants.
+func (e *SessionAccessEvaluator) FulfilledFor(participants []SessionAccessContext) (bool, PolicyOptions, error) {
+	if len(e.requires) == 0 || !e.isPoisoned() {
+		return true, PolicyOptions{}, nil
+	}
+
+	for _, requirePolicy := range e.requires {
+		left := requirePolicy.Count
+
+		for _, participant := range participants {
+			allowPolicies := getAllowPolicies(participant)
+			for _, allowPolicy := range allowPolicies {
+				matchesPredicate, err := e.matchesPredicate(&participant, requirePolicy, allowPolicy)
+				if err != nil {
+					return false, PolicyOptions{}, trace.Wrap(err)
+				}
+
+				if matchesPredicate && e.matchesJoin(allowPolicy) {
+					left--
+					break
+				}
+			}
+
+			if left <= 0 {
+				options := PolicyOptions{}
+
+				switch requirePolicy.OnLeave {
+				case types.OnSessionLeaveTerminate:
+					options.TerminateOnLeave = true
+				case types.OnSessionLeavePause:
+					options.TerminateOnLeave = false
+				default:
+					return false, PolicyOptions{}, trace.BadParameter("unsupported on_leave policy: %v", requirePolicy.OnLeave)
+				}
+
+				return true, options, nil
+			}
+		}
+	}
+
+	return false, PolicyOptions{}, nil
+}
+
+// isPoisoned checks if a set of roles contains any v5 roles.
+// If a set only has v4 or earlier roles, we don't want to apply the access checks
+// to ssh sessions.
```

Comment:
Updated the godoc. It's due to the fact that we need fallback behaviour to not break existing SSH session behaviour but we don't for k8s since it's multiparty feature is completely new.